Fix output path issues with customized builds

Nirmal4G opened the pull request:
This PR Fixes NuGet/Home#9234

 - This fixes a host of path related issues when doing a customized build with Pack targets.
 - It also obsoletes the `PackageOutputPath` in the SDK's `DefaultOutputPaths` targets.
 - It also ensures non-sdk style projects doesn't have to specify some default paths.

Requires microsoft/msbuild#5238

Nirmal4G committed:
Move output paths default initialization out of the target

This fixes a host of path related issues when doing a customized build with Pack targets.
It also obsoletes the 'PackageOutputPath' in the SDK's 'DefaultOutputPaths' targets.
It also ensures non-sdk style projects doesn't have to specify some default paths.

dominoFire commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -282,8 +284,8 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
-  
-  <Target Name="_GetProjectReferenceVersions" 
+
```
Remove the space comments, since it adds no value to the changeset.

Nirmal4G replied:
It's fine but every time I open/save the file, it removes the leading and tailing whitespaces...

Nirmal4G replied:
Will do clean up at the end.

Nirmal4G replied:
Done.

dominoFire commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -407,7 +409,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     </ItemGroup>
   </Target>
-  
+
```
Remove the space changes, since it adds no value to the changeset.

Nirmal4G replied:
Will do clean up at the end.

nkolev92 replied:
Sounds good, let us know when you are done so we can review again.

dominoFire commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -40,14 +40,20 @@ Copyright (c) .NET Foundation. All rights reserved.
     <AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder Condition="'$(SymbolPackageFormat)' == 'snupkg'">.pdb</AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>
     <SuppressDependenciesWhenPacking Condition="'$(SuppressDependenciesWhenPacking)' == ''">false</SuppressDependenciesWhenPacking>
   </PropertyGroup>
+
```
Add a test in [NuGet/NuGet.Client/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs) to verify the correct behavior.

You can find inspiration from similar tests. For example:

https://github.com/NuGet/NuGet.Client/blob/ca008ff53cd941d31cba2052c1fcf4c31770b294/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs#L4061-L4083

Nirmal4G replied:
OK. Will do.

nkolev92 replied:
I don't think that'd work.
The problem is that those tests use the SDK and the SDK is the one that sets defaults.

nkolev92 replied:
I'd look into https://github.com/NuGet/NuGet.Client/tree/dev/test/NuGet.Core.FuncTests/Msbuild.Integration.Test and see if it's at all possible there.

Nirmal4G replied:
> SDK is the one that sets defaults

Do we want the SDK to set the defaults? Both `RestoreOutputPath` and `NuSpecOutputPath` have defaults set in NuGet targets.

> I'd look into [MSBuild.Integration.Test](/NuGet/NuGet.Client/tree/dev/test/NuGet.Core.FuncTests/Msbuild.Integration.Test) and see if it's at all possible there.

There's a PR already on the SDK side. I have tested locally with only pack related tests. They all pass. I recommend to merge this first, since the SDK depends on it.

> I don't think that'd work.

I could add a simple test that fails if the defaults had not been set at the end of the first pass.

nkolev92 replied:
> Do we want the SDK to set the defaults? Both RestoreOutputPath and NuSpecOutputPath have defaults set in NuGet targets.

We don't need them to.
No one should really use RestoreOutputPath, it's not consistently interpreted, it's MSBuildProjectExtensionsPath that matters.

nkolev92 replied:
> There's a PR already on the SDK side. I have tested locally with only pack related tests. They all pass. I recommend to merge this first, since the SDK depends on it.

Yeah we'll merge this one first. It doesn't harm us to have the same logic in 2 places.

> I could add a simple test that fails if the defaults had not been set at the end of the first pass.

How would you achieve that if the targets set the defaults themselves?

Nirmal4G replied:
> We don't need them to

Thanks for clarification.

> How would you achieve that if the targets set the defaults themselves?

SDK tests reports error if the defaults not set. I have a project that repros the issue. I have to debug those to find the cause.

nkolev92 replied:
Just do a best effort and we'll try to help if you get stuck. :)

dominoFire reviewed (dismissed):
Thank you so much for your contribution. We need some tests for this change.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -40,14 +40,20 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <PropertyGroup>
+    <RestoreOutputPath Condition="'$(RestoreOutputPath)' == ''">$(MSBuildProjectExtensionsPath)</RestoreOutputPath>
+    <PackageOutputPath Condition="'$(PackageOutputPath)' == ''">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
+    <NuspecOutputPath Condition="'$(NuspecOutputPath)' == ''">$(BaseIntermediateOutputPath)$(Configuration)\</NuspecOutputPath>
```
I think both this and the above should be: OutputPath

nkolev92 replied:
Pack should be invoked without a target framework property, because it acts all on all build outputs so we want to be in the root folder.

Nirmal4G replied:
> it acts all on all build outputs

Including `Configuration` and `Platform`/`PlatformName`?

nkolev92 replied:
It's 1 package per project.

Nirmal4G replied:
Then `Base*OutputPath` properties are the best. We can remove `Configuration` too, as per your suggestion. This means we can't produce `Debug` and `Release` packages by default. That would be a breaking change...

nkolev92 replied:
OutputPath includes the configuration but not the platform. 

Keeping OutputPath preserves the current behavior.

Nirmal4G replied:
The current state of the `OutputPath` in the SDK.

```xml
    <BaseOutputPath Condition="'$(BaseOutputPath)' == ''">bin\</BaseOutputPath>
    <BaseOutputPath Condition="!HasTrailingSlash('$(BaseOutputPath)')">$(BaseOutputPath)\</BaseOutputPath>
    <OutputPath Condition="'$(OutputPath)' == '' and '$(PlatformName)' == 'AnyCPU'">$(BaseOutputPath)$(Configuration)\</OutputPath>
    <OutputPath Condition="'$(OutputPath)' == '' and '$(PlatformName)' != 'AnyCPU'">$(BaseOutputPath)$(PlatformName)\$(Configuration)\</OutputPath>
    <OutputPath Condition="!HasTrailingSlash('$(OutputPath)')">$(OutputPath)\</OutputPath>
```

See, `PlatformName` (derived from `Platform`) is being added to the path.

nkolev92 replied:
Nice catch. I only tested with AnyCPU. 

The way I see it, using OutputPath is a consistent behavior. I don't want us to regress something, while we are trying to improve the defaults.

Nirmal4G replied:
Everything we tested for is `bin\Debug` and nothing else. This actually fixes the problem before we even encounter them... I've already tested with my local projects. There **will** be **no** problems.

If it were me, I would also vote to remove that `Configuration` part too.

nkolev92 replied:
I understand that we have preferences, but in this PR we are merely adjusting the defaults :) 
I want to keep the scope to that.

Nirmal4G replied:
OK.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -40,14 +40,20 @@ Copyright (c) .NET Foundation. All rights reserved.
   </PropertyGroup>
+
+  <PropertyGroup>
+    <RestoreOutputPath Condition="'$(RestoreOutputPath)' == ''">$(MSBuildProjectExtensionsPath)</RestoreOutputPath>
```
Why do you need this set globally?

Nirmal4G replied:
We don't need to.

This way, these properties are set before the second pass and before any target gets called. So, we can use these on item groups in a customized build.

But the downside is, it allows anyone to unset those, if they did it after the default targets. If we don't want that to happen, we could keep the same duplicated logic inside that target too. but I doubt that anyone would want to do that...

nkolev92 replied:
The NuGet.targets already set the RestoreOutputPath. 
They can't really work without one another so as such I'd just revert this part of the change.

Nirmal4G replied:
> The NuGet targets already set the `RestoreOutputPath`

in the Target here...

https://github.com/NuGet/NuGet.Client/blob/19aec47a81889a9bed937c404d6e11b0701bf086/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L643-L654

Pack targets doesn't depend on this target to get `RestoreOutputPath`.

nkolev92 replied:
I see. 

Pack can just depend on MSBuildProjectExtensionsPath then. No need to reuse a property that's not consistently used.

Nirmal4G replied:
I will remove it from global declaration.

> Pack can just depend on MSBuildProjectExtensionsPath then

Can I lock the property on both Pack and Restore targets?

nkolev92 replied:
> Can I lock the property on both Pack and Restore targets?

Can you clarify this?

Nirmal4G replied:
Remove ` Condition=" '$(RestoreOutputPath)' == '' " ` and set `TreatAsLocalProperty`.

Nirmal4G replied:
There are only two instances on SDK where the property is used. I can add changes to the existing PR if it's okay with you.

Nirmal4G replied:
Remove ` Condition=" '$(RestoreOutputPath)' == '' " ` and set `TreatAsLocalProperty`.

There are only two instances (`ComposeStore.targets` and `CrossGen.targets`) on SDK where the property is used.

Can you confirm if they have a specific reason for this?

If they have a reason, then we can't make it a local property.

nkolev92 replied:
I wouldn't go beyond ensuring that there's a sensible default equivalent to what the SDK was setting.

Nirmal4G replied:
Then I'll go ahead with the changes. Let us test first, then I'll revert any undesired ones.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -63,16 +63,47 @@ Copyright (c) .NET Foundation. All rights reserved.
+      <NuspecOutputPath Condition="'$(NuspecOutputPath)' == ''">$(BaseIntermediateOutputPath)</NuspecOutputPath>
+    </PropertyGroup>
 
+    <ConvertToAbsolutePath Paths="$(RestoreOutputPath)">
```
Looks like the changeset has gotten very large right now. 
I would avoid refactoring too many things.
Please remove the white space changes.

Nirmal4G replied:
Done.

Nirmal4G commented:
@dominoFire Can you release the PR, so that I can test my changes. My machine is not that powerful.

Nirmal4G commented:
I don't know how to fix the whitespace changes permanently. Both VS and VS Code formats them everytime I save. VS is the worst, It formatted the XML heavily. I had to undo most changes by hand. I don't want to change my settings in either editor, so, I disabled XML tools in code and using it instead of VS. I also did it via GitHub, but introduces new spaces instead. I'll have to fix them manually via notepad at the end of the PR.

**So Sorry**

Nirmal4G commented:
I have been going through the tests. Why are all tests written with project directory as the `PackageOutputPath` and only `obj` as `NuSpecOutputPath` instead of `bin\{<Platform>\}Debug` and `obj\Debug` as these were previous defaults?

Is it okay if I left them as it is?

The reason all the tests pass because all of them overrides the output paths from the command line invocation. So, No matter what we set they'll all pass. But when I rewrote the logic for invoking Pack cmd using the defaults, they all failed since they were hard coded for the above mentioned logic.

Do I fix all of the tests or add a default path test separately while keeping the others working?

nkolev92 commented:
> I have been going through the tests. Why are all tests written with project directory as the PackageOutputPath and only obj as NuSpecOutputPath instead of bin\{<Platform>\}Debug and obj\Debug as these were previous defaults?

I don't think there's a super special reason for this. It's all at the mercy of the person writing the tests. 

I'd don't change the tests, they are a good test that the packageoutputpath behavior is preserved.

nkolev92 reviewed (commented):
There are lots of formatting changes in the targets file. 

The way I see it, I don't tihnk this change set needs to be bigger than 10 lines. 

Can you please fix the formatting changes without a functional impact. 

Thanks!

nkolev92 commented on .editorconfig:
```diff
@@ -10,7 +10,7 @@ root = true
 ; (Please don't specify an indent_size here; that has too many unintended consequences.)
 charset = utf-8
-trim_trailing_whitespace = true
+trim_trailing_whitespace = false
```
Please revert this.

Nirmal4G replied:
VS and VS Code reads this and automatically removes trailing whitespace. If I undo this change, the files that I'll edit after the revert will still have trailing whitespace removed.

Help me here...

nkolev92 replied:
it should only do that if you select the lines. 
Additionally you can do all your changes with this setting and then before commiting (if you do it from the commandline), just revert this file back to the dev one. 

> git checkout dev -- .editorconfig

should do it.

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:
```diff
@@ -5848,7 +5848,6 @@ public void PackCommand_ProjectFile_PackageIconUrl_WithNuspec_WithPackTask_Warns
     <OutputPath>out</OutputPath>
     <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
     <PackageIconUrl>https://test/icon.jpg</PackageIconUrl>
-    <PackageOutputPath>bin\Debug\</PackageOutputPath>
```
Why did you need to remove this?

Nirmal4G replied:
That's actually the default now, as it always been, so don't need this.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -63,16 +62,46 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <!--
+    ============================================================
+    _GetAbsoluteOutputPathsForPack
+    Gets the Absolute output paths for Pack.
```
```suggestion
    Gets the absolute output paths for Pack.
```

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -63,16 +62,46 @@ Copyright (c) .NET Foundation. All rights reserved.
-  <Target Name="_GetOutputItemsFromPack"
-          Returns="@(_OutputPackItems)">
+  <PropertyGroup>
+    <PackageOutputPath Condition="'$(PackageOutputPath)' == ''">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
```
```suggestion
    <PackageOutputPath Condition="'$(PackageOutputPath)' == ''">$(OutputPath)</PackageOutputPath>
```

Nirmal4G replied:
This **change** is the **actual PR**.

nkolev92 replied:
The change is scope of the PackageOutputPath property, so that installing the pack sdk into non-sdk PackageReference based projects enables to `msbuild /t:pack` and not get the failure about a missing PackageOutputPath.

Nirmal4G replied:
I just tested on non-style projects, it doesn't work since `BaseOutputPath` is not there in common targets. Either we can change to `OutputPath` when base is not available or we could add base to common targets.

zivkan replied:
I don't have an answer, but I wanted to thank you for testing on non-SDK style projects. This isn't the easiest change to test, so I really appreciate you made the effort of testing different scenarios.

Nirmal4G replied:
```suggestion
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(BaseOutputPath)' == ''">$(OutputPath)</PackageOutputPath>
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(BaseOutputPath)' != ''">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
```

Nirmal4G replied:
Well, I do have a **lot** of framework projects!! 😎

Nirmal4G replied:
How about...

This?
```suggestion
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(BaseOutputPath)' == ''">$(OutputPath)</PackageOutputPath>
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(BaseOutputPath)' != ''">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
```

OR This?
```suggestion
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(UsingMicrosoftNETSdk)' != 'true'">$(OutputPath)</PackageOutputPath>
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(UsingMicrosoftNETSdk)' == 'true'">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
```

I prefer the first one, since it depends on the property itself. The second one will work (even from v1) but what if they use a custom Sdk (I do) with similar defaults as .NET Sdk, then the latter fails.

nkolev92 replied:
Apologies for the delay. 

I am not sure we are all on the same page on the problem we are trying to fix. 

I have replied to the original issue because I think the discussion is best had there. https://github.com/NuGet/Home/issues/9234#issuecomment-604114471

Specifically, I cannot accept this change because in non-SDK based projects it will end up a directory structure as follows: 

App\
  bin\Debug\*.dll, *.exe
  Debug\*.nupkg
  obj\Debug\*.nuspec, *.dll

In SDK based projects the structure is and should remain:

bin\Debug\ *.nupkg
bin\Debug\targetFrameworkName1\ *.dll
bin\Debug\targetFrameworkName2\ *.dll
obj\Debug\ *.nuspec
obj\Debug\targetFrameworkName1\ *.dll
obj\Debug\targetFrameworkName2\ *.dll

The outputs should mimic one another.

Nirmal4G replied:
> Debug*.nupkg

This won't happen in either of those. package will always end up in right location.

https://github.com/NuGet/Home/issues/9234#issuecomment-604138764

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -142,27 +171,12 @@ Copyright (c) .NET Foundation. All rights reserved.
 
-  <Target Name="_CalculateInputsOutputsForPack" DependsOnTargets="_GetOutputItemsFromPack">
+  <Target Name="_CalculateInputsOutputsForPack"
+          DependsOnTargets="_GetAbsoluteOutputPathsForPack;_GetOutputItemsFromPack">
```
You don't need this change.
_GetOutputItemsFromPack already depends on _GetAbsoluteOutputPathsForPack.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -192,8 +206,15 @@ Copyright (c) .NET Foundation. All rights reserved.
 
-  <Target Name="GenerateNuspec" DependsOnTargets="$(GenerateNuspecDependsOn);_CalculateInputsOutputsForPack;_GetProjectReferenceVersions;_InitializeNuspecRepositoryInformationProperties" Condition="$(IsPackable) == 'true'"
-          Inputs="@(NuGetPackInput)" Outputs="@(NuGetPackOutput)">
+  <Target Name="GenerateNuspec"
```
Is there a functional change here?
I'd avoid this formatting change if not.

Nirmal4G replied:
Better readability, I guess.

nkolev92 replied:
If there's no functional change, I'd rather you keep the same as before. 
In targets preserving the history is helpful.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -646,7 +646,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     <!-- Determine the restore output path -->
     <PropertyGroup Condition=" '$(PackageReferenceCompatibleProjectStyle)' == 'true' OR '$(RestoreProjectStyle)' == 'ProjectJson' ">
-      <RestoreOutputPath Condition=" '$(RestoreOutputPath)' == '' " >$(MSBuildProjectExtensionsPath)</RestoreOutputPath>
+      <RestoreOutputPath>$(MSBuildProjectExtensionsPath)</RestoreOutputPath>
```
I wouldn't change this. 
We recommend people not use it, but we just don't know who we'd end up breaking.

Is there a functional value to this or just a cleanup?

Nirmal4G replied:
Changing it didn't break any tests, so I guess it's only cleanup but will need to talk to sdk team about this.

I'll separate this into a new PR, we can make it in for .NET 5.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs:
```diff
@@ -187,17 +187,24 @@ internal CommandRunnerResult RunDotnet(string workingDirectory, string args, boo
-            => PackProjectOrSolution(workingDirectory, $"{projectName}.csproj", args, nuspecOutputPath, validateSuccess);
+        {
+            // We can't provide empty or spaces as arguments if we used `string.IsNullOrEmpty` or `string.IsNullOrWhiteSpace`.
+            if (nuspecOutputPath != null) args = $"{args} /p:NuspecOutputPath={nuspecOutputPath}";
```
If you are gonna chnage it, use:

```suggestion
            if (nuspecOutputPath != null)
            {
                args = $"{args} /p:NuspecOutputPath={nuspecOutputPath}";
            }
```

See Rule 1 in https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#c-coding-style

Nirmal4G replied:
So sorry about that, it's a wrong patch.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs:
```diff
@@ -187,17 +187,24 @@ internal CommandRunnerResult RunDotnet(string workingDirectory, string args, boo
         internal CommandRunnerResult PackSolution(string workingDirectory, string solutionName, string args, string nuspecOutputPath = "obj", bool validateSuccess = true)
-            => PackProjectOrSolution(workingDirectory, $"{solutionName}.sln", args, nuspecOutputPath, validateSuccess);
+        {
+            if (nuspecOutputPath != null) args = $"{args} /p:NuspecOutputPath={nuspecOutputPath}";
```
Same as above.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -42,13 +42,23 @@ public void PackCommand_PackNewDefaultProject_NupkgExists()
                 Assert.True(File.Exists(nuspecPath), "The intermediate nuspec file is not in the expected place");
 
+                // with default output paths
+                msbuildFixture.PackProject(workingDirectory, projectName, string.Empty, null);
```
I would add a separate test for the default case.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -96,6 +108,16 @@ public void PackCommand_PackNewDefaultProject_IncludeSymbolsWithSnupkg()
                 Assert.True(File.Exists(nuspecPath), "The intermediate nuspec file is not in the expected place");
 
+                // with default output paths
+                msbuildFixture.PackProject(workingDirectory, projectName, $"--include-symbols /p:SymbolPackageFormat=snupkg", string.Empty);
```
Same here, please add a new test targeting that specific scenario.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -2649,16 +2671,25 @@ public void PackCommand_PackSolution_AddsProjectRefsAsPackageRefs()
 
-                // Act
-                msbuildFixture.PackSolution(testDirectory, solutionName, $"-o {testDirectory}");
+                // with custom output paths
```
This one seems like an odd test to change. 
Can you elaborate why?

Nirmal4G replied:
I already have separate tests, but I wanted to get your opinion on this patch.

I want to test the scenario where I already built the project or solution with the defaults and then without cleaning it up, pack again into new custom path or vice versa.

Nirmal4G replied:
I'll separate those out, thanks.

Nirmal4G commented:
Also, why the CI won't build every time I push. The label says, `Approved for CI`.

Nirmal4G committed:
Remove default value to PackageOutputPath

The Default is now implied and points to the same directory.
It is now redundant, and by removing it, the tests should not fail.

Nirmal4G committed:
Added new Tests to check default output paths

- Checks the default output paths from Project/Solution builds.
- Check both the default and custom paths from continuous Project/Solution builds.

nkolev92 commented:
1 label = 1 build. :)

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets:
```diff
@@ -63,8 +62,27 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <Target Name="_GetAbsoluteOutputPathsForPack">
+    <PropertyGroup>
+      <RestoreOutputPath Condition="'$(RestoreOutputPath)' == ''">$(MSBuildProjectExtensionsPath)</RestoreOutputPath>
+      <PackageOutputPath Condition="'$(PackageOutputPath)' == ''">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
```
Do we need to apply the default in both places? Or is 1 enoguh?

Nirmal4G replied:
within `_GetAbsoluteOutputPathsForPack` is enough but I'm afraid that we might regress something that haven't been properly tested, that's why I added them globally too! (as `NuSpecOutputPath` was already there).

Also one benefit (side effect) is that the defaults will be available after MSBuild's property pass.

**Can anyone clarify this?**

nkolev92 replied:
@rainersigwald 
Any gotchas with properties declared inside a target vs globally? 

Specifically any concerns with moving the output defaults from inside a target where they are used to global. 
The expectation is what the output defaults to is something that doesn't change. 

I *think* adding them globally is safe.

rainersigwald replied:
Properties defined inside targets don't get set until the target runs, which can present ordering issues. In this case, since they're unconditionally defaulted outside a target, it'd be pretty funky (though certainly possible) for them to be unset when the target started. A project would have to explicitly unset them either in a later .targets or in a target.

The biggest problem with leaving them only as evaluation-time targets is that they can be modified by user code. But that's also a configurability advantage--and this PR only sets defaults, not a problem, hopefully.

I lean toward "only outside the target".

Also, a big nit to pick with y'all :): [global properties](https://docs.microsoft.com/visualstudio/msbuild/msbuild-properties?view=vs-2019#global-properties) are a thing, but not something that's on the table here. You scared me at first

nkolev92 replied:
> > I lean toward "only outside the target".

Thanks for the explanation. @nirmal4g let's do outside of the target only.

> Also, a big nit to pick with y'all :): global properties are a thing, but not something that's on the table here. You scared me at first

😆 I appreciate the nit pick!
Life is simpler when we use the same vernacular 😄

Nirmal4G committed:
Remove duplicate logic inside the target

The duplicate logic is not needed.
(edited anyway but will push if anyone approves)

nkolev92 commented:
Hey @Nirmal4G are you still planning to do work here?

Nirmal4G commented:
Tests pending, requires `BaseOutputPath` support in MSBuild, PR is ongoing.

Nirmal4G commented:
@nkolev92 Tests pending, requires `BaseOutputPath` support in MSBuild, PR is ongoing.

nkolev92 commented:
@Nirmal4G 

We can't rely on that change unfortunately. 
In old style csproj projects, people install the pack package in projects that might not necessarily use the latest version of MSBuild. 
We can try to do a best effort in old style csproj or just do nothing. 

The MSBuild change is in a very sensitive area, so validating that scenarios are not gonna break is a very labor intensive task.

Nirmal4G commented:
@nkolev92 

> We can try to do a best effort in old style csproj

If we have `BaseOutputPath` set, we can take it as it is else we can set it to a default value `bin`. Then the logic flows as expected. Is that okay with you?

> The MSBuild change is in a very sensitive area, so validating that scenarios are not gonna break is a very labor intensive task.

After the msbuild PR is done, I'll make sure, that the necessary changes are reflected here.

zivkan commented:
> After the msbuild PR is done, I'll make sure, that the necessary changes are reflected here.

@nkolev92's point is that people using old version of MSBuild, for example, people using Visual Studio 2017, have PackageReference to `NuGet.Build.Tasks.Pack`. Even once your MSBuild PR is merged, Visual Studio 2017, and other old versions of msbuild, isn't going to get the change, so we can't use it in `NuGet.Build.Tasks.Pack`.

Nirmal4G commented:
@nkolev92 @zivkan _Am I that hard to understand or is it something I missed_?

> projects that might not necessarily use the latest version of MSBuild

That I get. I was offering up a solution for those scenarios. `OutputPath` or `bin\` as a fallback.

```xml

  <PropertyGroup Condition="'$(Variant)' == 'v1'">
    <BaseOutputPath Condition="'$(BaseOutputPath)' == ''">bin\</BaseOutputPath>
    <PackageOutputPath Condition="'$(PackageOutputPath)' == ''">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Variant)' == 'v2'">
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(BaseOutputPath)' == ''">$(OutputPath)</PackageOutputPath>
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(BaseOutputPath)' != ''">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Variant)' == 'v3'">
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(UsingMicrosoftNETSdk)' != 'true'">$(OutputPath)</PackageOutputPath>
    <PackageOutputPath Condition="'$(PackageOutputPath)' == '' AND '$(UsingMicrosoftNETSdk)' == 'true'">$(BaseOutputPath)$(Configuration)\</PackageOutputPath>
  </PropertyGroup>

```

All of these are the workarounds, I thought of when solving the issue. If any of these can be applied, please let me know.

nkolev92 commented:
I think we were leaning toward the 3rd variant (based on our discussions in this PR and issue).

The logic being that for SDK projects, the change is backwards compatible. 
For non-sdk projects, it's a best effort to use the output path, because the baseoutputpath is not applied consistently.

Nirmal4G commented:
@nkolev92 

> For non-sdk projects, it's a best effort to use the output path, because the baseoutputpath is not applied consistently.

_True_. But keep in mind that projects which use custom SDKs (that don't declare `UsingMicrosoftNETSdk`) will use `OutputPath` even if they've `BaseOutputPath` defined.
Isn't it better to check for `BaseOutputPath` itself than to check with a property that only lives in Microsoft.NET.Sdk. This way, when `BaseOutputPath` support is added to common targets, it'll pickup on that (future-proof).

It's a win-win for both types of projects! With that, I lean towards **Variant 2**.

nkolev92 commented:
Good point, let's do 2. 

Sorry, we've been all over the place in this PR/issue.

Nirmal4G commented:
> Good point, let's do 2.

Done. Pack task tests are **missing** for non-sdk style projects. Should I've to create a **separate PR** or **do it here** for these missing tests?

> Sorry, we've been all over the place in this PR/issue.

It's OK. I don't mind.

- Only complain is that the repo is hard to understand/navigate. Could be much simpler.
- Building in VS IDE does not work properly. I followed the instructions in the guidelines but 2 out of 3 times, it does not work.
- PR builds could be made public in Azure pipelines. Helps us understand build failures

@dotnet repos use [Arcade](/dotnet/arcade) Sdk which is better in so many ways. I wish, you guys would use that _soon_.

nkolev92 commented:
> Only complain is that the repo is hard to understand/navigate. Could be much simpler.

We're trying to make it more contributor friendly. The complexity is a consequence of NuGet being a part of 3 official products actively maintained in this repo. 

> Building in VS IDE does not work properly. I followed the instructions in the guidelines but 2 out of 3 times, it does not work.

Things should pretty much work in VS. Were you facing issues running the PackCommandTests?
I think the integration tests for nuget.exe and dotnet.exe pack are the ones that I can reason through a failure. We'd love to learn more. Follow up https://github.com/NuGet/Home/issues/9503.

> PR builds could be made public in Azure pipelines. Helps us understand build failures

We're not there yet, but we'd like to get there soon.

Nirmal4G commented:
> Follow up [NuGet/Home#9503](/NuGet/Home/issues/9503)

The page says `This issue has been moved to a repository you don't have access to.`

**All Done in this PR**. Please let me know, if there are any test failures.

Nirmal4G commented:
@nkolev92 🔔

Nirmal4G commented:
@dominoFire @nkolev92 🔔

Nirmal4G commented:
@dominoFire @nkolev92 🔔

Nirmal4G commented:
@dominoFire @nkolev92 Any status?

Nirmal4G commented:
@dominoFire @nkolev92 Can we check this and merge this in? It's been too long.

Nirmal4G commented:
@dominoFire @nkolev92 🔔

Nirmal4G commented:
@dominoFire @nkolev92 🔔

nkolev92 reviewed (approved):
Sorry for the delay @Nirmal4G. 

Thank you for all the hard work on this issue.

nkolev92 commented:
@Nirmal4G I kicked off the build.

Nirmal4G commented:
@nkolev92 Seems some tests with .NET Core are failing... Can you share the report? Thanks.

Nirmal4G commented:
@nkolev92

nkolev92 commented:
I reran the failed jobs.

Nirmal4G commented:
@nkolev92 Still is failing, is it the CI issue or the actual tests have been failing?

Nirmal4G commented:
@nkolev92 seems like a CI issue.

Nirmal4G commented:
@nkolev92 Functional Tests with .NET Core is failing. It says "failed to run tests", so... May be CI issues?

nkolev92 commented:
@Nirmal4G A bunch of tests are failing: 

* PackCommand_NewProject_ContinuousOutputInBothDefaultAndCustomPaths
* PackCommand_NewProject_OutputsInDefaultPaths
* DotnetToolTests_ToolPackageAndPlatformsPackageAnyRID_SucceedsAsync
* DotnetToolTests_AutoreferencedDependencyRegularDependencyAndToolPackagToolRestore_ThrowsAsync
* DotnetToolTests_ToolPackageAndPlatformsPackageAnyRID_SucceedsAsync\
* DotnetToolTests_AutoreferencedDependencyAndToolPackagToolRestore_SucceedsAsync
* DotnetToolTests_AutoreferencedDependencyRegularDependencyAndToolPackagToolRestore_ThrowsAsync
* DotnetToolTests_AutoreferencedDependencyAndToolPackagToolRestore_SucceedsAsync
* DotnetToolTests_IncompatibleAutorefPackageAndToolsPackageAsync
* DotnetToolTests_IncompatibleAutorefPackageAndToolsPackageAsync
* PackCommand_NewProject_ContinuousOutputInBothDefaultAndCustomPaths
* PackCommand_NewProject_OutputsInDefaultPaths
* PackCommand_NewSolution_OutputInDefaultPaths
* PackCommand_NewSolution_ContinuousOutputInBothDefaultAndCustomPaths

Nirmal4G commented:
@nkolev92 I have fixed the tests. Please do run the CI again.

nkolev92 commented:
@Nirmal4G Can you rebase? 

Are the common.ps1 changes intentional?

Nirmal4G commented:
@nkolev92 

> Can you rebase?

Sure. Done.

> Are the common.ps1 changes intentional?

Yes, It fixed some of the issues when building locally.

Nirmal4G commented:
🔔 CI Please

Nirmal4G commented:
@nkolev92 @zivkan

Nirmal4G commented:
🔔 CI Please

Nirmal4G commented:
@nkolev92

Nirmal4G commented:
🔔 I think I fixed the script... Can you start CI again?

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs:
```diff
@@ -153,7 +153,12 @@ public LockFile ExistingLockFile
         /// <summary>
         /// MSBuildProjectExtensionsPath
         /// </summary>
-        public string MSBuildProjectExtensionsPath { get; set; }
```
I'm really not sure why these changes are being made. 

This keeps increasing the scope of the PR.

Nirmal4G replied:
I thought it was a leftover from when you guys combined the `BaseIntermediateOutputPath`, `MSBuildProjectExtensionsPath` into `RestoreOutputPath`. I stumbled upon these when I was trying to fix some tests. So, I corrected it!! may be?

Can I separate those into a new PR, then?

nkolev92 replied:
Stuff in src\NuGet.Core is part of a public API. 
We're a tool first, but there are customer that use our packages as libraries, so we've strayed away from making refactoring changes.

nkolev92 replied:
We added this doc recently, https://github.com/NuGet/NuGet.Client/blob/dev/docs/nuget-sdk.md

Nirmal4G replied:
Looks like Project Json uses it. So, I have reverted this particular change. Thanks for the heads up!!

nkolev92 commented on build/common.ps1:
```diff
@@ -157,95 +157,81 @@ Function Install-DotnetCLI {
-    $MSBuildExe = Get-MSBuildExe $vsMajorVersion
-    $CliBranchListForTesting = & $msbuildExe $NuGetClientRoot\build\config.props /v:m /nologo /t:GetCliBranchForTesting
-    $CliBranchList = $CliBranchListForTesting.Split(';');
+    $env:DOTNET_HOME = $CLIRoot
```
Can you elaborate what's the issue with the scripts that you are trying to fix?

Nirmal4G replied:
I mentioned the issues in the commit's description.

- MSBuild picks up response files automatically, which messes up reading from standard output.
- The overridden version won't run the tests successfully. Since, some tests error out by saying 'Pkcs' dll missing.
- Cleanup the logic to make it better understandable.

nkolev92 replied:
> Since, some tests error out by saying 'Pkcs' dll missing.

That shouldn't be happening.

We test both 3.x and 5.x SDKs, that's why we download both

Nirmal4G replied:
Moved to #3623

Nirmal4G commented:
🔔 CI Please

nkolev92 commented on src/NuGet.Core/NuGet.Commands/MSBuildPackTargetArgs.cs:
```diff
@@ -15,6 +15,7 @@ public class MSBuildPackTargetArgs
         public HashSet<string> AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder { get; set; }
         public string AssemblyName { get; set; }
         public string NuspecOutputPath { get; set; }
+        public string PackageOutputPath { get; set; }
```
Why do we need this? The output path is already provided by other means.

Nirmal4G replied:
What other means and Where?

nkolev92 replied:
Redirect the packageoutputpath works in some capacity today right? 

https://github.com/NuGet/NuGet.Client/blob/1fcb0dc2507330453e41bc54a917e20356a85055/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTask.cs#L42

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:
```diff
@@ -181,8 +181,8 @@ private static ExternalProjectReference GetExternalProject(PackageSpec rootProje
                 RestoreOutputPath = project.PackageSpec.RestoreMetadata.ProjectStyle == ProjectStyle.ProjectJson ? rootPath : project.PackageSpec.RestoreMetadata.OutputPath,
-                DependencyGraphSpec = projectDgSpec,
                 MSBuildProjectExtensionsPath = projectPackageSpec.RestoreMetadata.OutputPath,
+                DependencyGraphSpec = projectDgSpec,
```
No functional change here, please revert.

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs:
```diff
@@ -30,7 +30,8 @@ internal static bool IsNoOpSupported(RestoreRequest request)
 
         /// <summary>
-        /// The cache file path is $(MSBuildProjectExtensionsPath)\$(project).nuget.cache
+        /// The cache file path will be $(RestoreOutputPath)\$(project).nuget.cache
```
We prefer not to use RestoreOutputPath, so I'd revert this. 
RestoreOutputPath was added in the past, but we've moved away from that because RestoreOutputPath would lead to the *build* not importing the `nuget.g.*`. 

We even added a warning for that: https://github.com/NuGet/NuGet.Client/commit/deab84db30cbd8cc738cd46ad694aebdbb43a1cf
```suggestion
        /// The cache file path will be $(RestoreOutputPath)\$(project).nuget.cache
```

Nirmal4G replied:
Please correct me if I'm wrong. The `RestoreOutputPath` is used to define where the restore cache files and the generated props/targets should go and it defaults to `MSBuildProjectExtensionsPath`, Right?

nkolev92 replied:
There are supposed to be the same. 

NuGet was defaulting to BaseIntermediateOutputPath, MSBuild was using MSBuildProjectExtensionsPath. These led to silent issues. 
At the same time, NuGet defined RestoreOutputPath and related which compounded the problem caused by 2 components using different props. 

The *only* right answer for end consumers is to use `MSBuildProjectExtensionsPath`.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/ProjectStyle.cs:
```diff
@@ -16,7 +16,7 @@ public enum ProjectStyle : ushort
 
         /// <summary>
-        /// MSBuild style, project.assets.json is generated in the RestoreOutputPath folder
+        /// MSBuild style, project.assets.json is generated in the RestoreOutputPath (aka MSBuildProjectExtensionsPath) folder
```
If you are changing it, it'd be better if we just say MSBuildProjectExtensionsPath.

Nirmal4G replied:
Response: https://github.com/NuGet/NuGet.Client/pull/3270#discussion_r465657566

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetToolTests.cs:
```diff
@@ -675,7 +674,7 @@ public async Task DotnetToolTests_IncompatibleAutorefPackageAndToolsPackageAsync
                 _msbuildFixture.CreateDotnetToolProject(solutionRoot: testDirectory.Path,
                     projectName: projectName, targetFramework: tfm, rid: rid,
-                    source: source, packages: packages);
+                    packageSource: source, packages: packages);
```
All these package source change seem unnecessary. 

I'm not sure the extra qualifier adds much to the method.

Nirmal4G replied:
Just making it clear what `source` means here. It's public method that we are using it to write tests. So, it should be clear on what `source` is.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs:
```diff
@@ -114,7 +253,7 @@ private static void CopyFromTemplate(string projectName, string args, string wor
         }
 
-        internal void CreateDotnetToolProject(string solutionRoot, string projectName, string targetFramework, string rid, string source, IList<PackageIdentity> packages, int timeOut = 60000)
+        internal void CreateDotnetToolProject(string solutionRoot, string projectName, string targetFramework, string rid, string packageSource = null, IList<PackageIdentity> packages = null, int timeOut = 60000)
```
No need to add `source` to `packageSource`. It increases the footprint of the change without adding much in the clarity department.

Nirmal4G replied:
Just making it clear what `source` means here. It's public method that we are using it to write tests. So, it should be clear on what `source` is.

nkolev92 reviewed (commented):
Hey @Nirmal4G,

We really appreciate all the effort you have put into this PR, especially your analysis which has been extremely detailed and spans more than just NuGet.

One of the things that would really help us expedite code reviews and the validation process is keeping PRs small and focused on solving one problem at the time. Therefore, we’d prefer it if you could move changes unrelated to the pack code in question to separate PRs as appropriate:

o Test code refactoring  
o Build script improvements.

Doing this will also ensure that we can identify and revert potential regressions quickly without pulling out all the improvements that you have been kindly working on.

Again, on behalf of the @NuGet/nuget-client team thank you for your contributions, and we are looking forward to working with you on many more.

Nirmal4G commented:
@nkolev92 CI Please.

Nirmal4G commented:
I have simplified this PR to only contain the `PackageOutputPath` changes only for SDK style projects. The testing for non-SDK style projects has been separated into a new PR which I'll open once this is merged.

nkolev92 commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -28,6 +28,194 @@ public PackCommandTests(MsbuildIntegrationTestFixture fixture)
+        }
+
+        [PlatformFact(Platform.Windows)]
+        public void PackCommand_NewSolution_ContinuousOutputInBothDefaultAndCustomPaths()
```
nit: The `In both` part confused me, made me think it was outputing in both location, which is obviously not something we're doing :)

Nirmal4G replied:
It is output'ing in both locations but not at the same time. It's like this, when you have a project, you build with defaults to see it succeed and then you run an official build to publish the artifacts in a different folder using different config without affecting the previous build. A common dev scenario (at least, in our case!!).

The name is little confusing, I agree but if you have an alternative, we could use that instead. I'll queue up the changes in the next PR if it's alright with you!?

nkolev92 replied:
I don't have a great suggestion, I'm fine with leaving it be.

nkolev92 reviewed (approved):
Thanks @Nirmal4G

Nirmal4G commented:
@nkolev92 All green, good to go!

nkolev92 commented:
@dominoFire your review is blocking the PR, can you please take a look. 

You can dismiss it if you'd rather not re-review.
